src/pathfinder/water_regions.cpp

```diff
@@ -0,0 +1,405 @@
+/*
+ * This file is part of OpenTTD.
+ * OpenTTD is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version 2.
+ * OpenTTD is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details. You should have received a copy of the GNU General Public License along with OpenTTD. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+ /** @file water_regions.h Handles dividing the water in the map into square regions to assist pathfinding. */
+
+#include "stdafx.h"
+#include "map_func.h"
+#include "water_regions.h"
+#include "map_func.h"
+#include "tilearea_type.h"
+#include "track_func.h"
+#include "transport_type.h"
+#include "landscape.h"
+#include "tunnelbridge_map.h"
+#include "follow_track.hpp"
+#include "ship.h"
+#include "viewport_func.h" // For debug drawing
+
+#include <vector>
+#include <array>
+#include <unordered_set>
+#include <unordered_map>
```

Comment:
`unorderd_set`/`unordered_map` seem a bit tricky in the pathfinder, as they can yield different order of elements on different platforms. So, that's something to take into account when testing/reviewing.